include/envoy/server/http_route_manager.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "envoy/router/rds.h"
+
+namespace Envoy {
+namespace Server {
+
+class HttpRouteManager {
+public:
+  virtual ~HttpRouteManager() {}
+
+  /**
+   * @return std::vector<Router::RouteConfigProviderSharedPtr> a list of all the
+   * RouteConfigProviders currently loaded.
+   */
+  virtual std::vector<Router::RouteConfigProviderSharedPtr> routeConfigProviders() PURE;
+
+  /**
+   * Add a RouteConfigProvider. Ownership of the RouteConfigProvider is shared by all the
+   * ConnectionManagers who own a RouteConfigProviderSharedPtr. The HttpRouteManager holds weak_ptrs
+   * to the RouteConfigProviders.
+   * @param name supplies the name of the RouteConfigProvider. This is a unique identifier.
+   * @param routeConfigProvider is a weak_ptr to the RouteConfigProvider to add.
+   */
+  virtual void
+  addRouteConfigProvider(const std::string& name,
```

Comment:
add/get/remove should be merged into a single API so that RAII can be used when the returned provider is destroyed. Basically, I would recommend doing the following:

- Return a shared_ptr to a route provider (or interface handle to the route provider).
- Internally inside the manager, store *both* a weak_ptr and a raw pointer. The raw pointer can be used to lookup the manager entry to remove when the last shared_ptr reference goes away. The weak_ptr can be used to create a new shared_ptr to return if needed (AFAIK there is no way to get the raw pointer out of a weak_ptr making this necessary).
- This will allow you to have a single get() method exposed.